frontend/src/metabase/selectors/metadata.js

```diff
@@ -123,22 +125,69 @@ export const getSegments = createSelector(
 );
 
-// MISC
-
-export const getParameterFieldValues = (state, props) => {
-    const fieldValues = getFieldValues(getIn(state, ["metadata", "fields", props.parameter.field_id]));
+// FIELD VALUES FOR DASHBOARD FILTERS / SQL QUESTION PARAMETERS
+
+// Returns a dictionary of field id:s mapped to matching field values
+// Currently this assumes that you are passing the props of <ParameterValueWidget> which contain the
+// `field_ids` array inside `parameter` prop.
+const getParameterFieldValuesByFieldId = (state, props) => {
+    // NOTE Atte Keinänen 9/14/17: Reading the state directly instead of using `getFields` selector
+    // because `getMetadata` doesn't currently work with fields of public dashboards
+    return _.chain(getIn(state, ["metadata", "fields"]))
+        .pick(getFields(state), ...props.parameter.field_ids)
+        .mapObject(getFieldValues)
+        .value()
+}
 
-    // HACK Atte Keinänen 7/27/17: Currently the field value analysis code only returns a single value for booleans,
-    // this will be addressed in analysis sync refactor
+// Custom equality selector for checking if two field value dictionaries contain same fields and field values
+// Currently we simply check if fields match and the lengths of field value arrays are equal which makes the comparison fast
+// See https://github.com/reactjs/reselect#customize-equalitycheck-for-defaultmemoize
+const createFieldValuesEqualSelector = createSelectorCreator(defaultMemoize, (a, b) => {
+// TODO: Why can't we use plain shallowEqual, i.e. why the field value arrays change very often?
+    return shallowEqual(_.mapObject(a, (values) => values.length), _.mapObject(b, (values) => values.length));
+})
+
+// HACK Atte Keinänen 7/27/17: Currently the field value analysis code only returns a single value for booleans,
+// this will be addressed in analysis sync refactor
+const patchBooleanFieldValues_HACK = (valueArray) => {
     const isBooleanFieldValues =
-        fieldValues && fieldValues.length === 1 && fieldValues[0] && typeof(fieldValues[0][0]) === "boolean"
+        valueArray && valueArray.length === 1 && valueArray[0] && typeof(valueArray[0][0]) === "boolean"
+
     if (isBooleanFieldValues) {
         return [[true], [false]];
     } else {
-        return fieldValues;
+        return valueArray;
     }
 }
 
+// Merges the field values of fields linked to a parameter and removes duplicates
+// We want that we have a distinct selector for each field id combination, and for that reason
+// we export a method that creates a new selector; see
+// https://github.com/reactjs/reselect#sharing-selectors-with-props-across-multiple-components
+// TODO Atte Keinänen 7/20/17: Should we have any thresholds if the count of field values is high or we have many (>2?) fields?
+export const makeGetMergedParameterFieldValues = () => {
+    return createFieldValuesEqualSelector(getParameterFieldValuesByFieldId, (fieldValues) => {
+        const fieldIds = Object.keys(fieldValues)
```

Comment:
I think using "fieldIds" here is confusing. For the longest time I thought these were literal field IDs (eg, `123`) rather than field values that have not been remapped.

Reply:
they are fieldIds – your confusion probably originates from terminology ambiguity between "parameter mapping" and "remapping" which are two different things ("parameter mapping" refers to how dashcards map to filters whereas "remapping" means replacing shown field values with something new)

Will try to clarify that in source code